# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/webidl/AudioData.webidl
# Commit: 745df1a31541
# Full Hash: 745df1a31541f1f0b6e31f1b8d7414bdbb2f1e83
# Author: Paul Adenot <paul@paul.cx>
# Date: 2024-03-06 04:41:40
# Regressor Bug: 1858958
# File Overlap Count: 1
# Description:
#   Bug 1858958 - WebIDL file and skeleton code for AudioData. r=chunmin,webidl,saschanaz
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D190943
# ==============================================================================

diff -r adf17fc47c84 -r 745df1a31541 dom/webidl/AudioData.webidl
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/dom/webidl/AudioData.webidl	Tue Mar 05 15:52:11 2024 +0000
@@ -0,0 +1,63 @@
+/* -*- Mode: IDL; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ * The origin of this IDL file is
+ * https://w3c.github.io/webcodecs/#audiodata
+ */
+
+// [Serializable, Transferable] are implemented without adding attributes here,
+// but directly with {Read,Write}StructuredClone and Transfer/FromTransfered.
+[Exposed=(Window,DedicatedWorker)]
+interface AudioData {
+  [Throws]
+  constructor(AudioDataInit init);
+
+  readonly attribute AudioSampleFormat? format;
+  readonly attribute float sampleRate;
+  readonly attribute unsigned long numberOfFrames;
+  readonly attribute unsigned long numberOfChannels;
+  readonly attribute unsigned long long duration;  // microseconds
+  readonly attribute long long timestamp;          // microseconds
+
+  [Throws]
+  unsigned long allocationSize(AudioDataCopyToOptions options);
+  [Throws]
+  undefined copyTo(
+      // bug 1696216: Should be `copyTo(AllowSharedBufferSource destination, ...)`
+      ([AllowShared] ArrayBufferView or [AllowShared] ArrayBuffer) destination,
+       AudioDataCopyToOptions options);
+  [Throws]
+  AudioData clone();
+  undefined close();
+};
+
+dictionary AudioDataInit {
+  required AudioSampleFormat format;
+  required float sampleRate;
+  required [EnforceRange] unsigned long numberOfFrames;
+  required [EnforceRange] unsigned long numberOfChannels;
+  required [EnforceRange] long long timestamp;  // microseconds
+  // bug 1696216: Should be AllowSharedBufferSource
+  required ([AllowShared] ArrayBufferView or [AllowShared] ArrayBuffer) data;
+  sequence<ArrayBuffer> transfer = [];
+};
+
+enum AudioSampleFormat {
+  "u8",
+  "s16",
+  "s32",
+  "f32",
+  "u8-planar",
+  "s16-planar",
+  "s32-planar",
+  "f32-planar",
+};
+
+dictionary AudioDataCopyToOptions {
+  required [EnforceRange] unsigned long planeIndex;
+  [EnforceRange] unsigned long frameOffset = 0;
+  [EnforceRange] unsigned long frameCount;
+  AudioSampleFormat format;
+};